src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1226 @@
+
+        *p->remaining -= 1;
+        if (*p->remaining == 0)
+                /* We want to crash the program if we can't exit the loop
+                 * cleanly, otherwise it will just hang */
+                assert_se(sd_event_exit(p->event, 0) >= 0);
+
+        free(p->job_path);
+
+        sd_event_source_disable_unref(p->job_interrupt_source);
+        sd_bus_slot_unref(p->job_properties_slot);
+        sd_bus_slot_unref(p->job_finished_slot);
+
+        return mfree(p);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(Userdata*, userdata_free);
+
+static Userdata* userdata_new(
+                void *userdata,
+                sd_bus *bus,
+                unsigned *remaining,
+                const char *target_path,
+                const char *target_id) {
+
+        _cleanup_(userdata_freep) Userdata *u = NULL;
+
+        u = new(Userdata, 1);
+        if (!u)
+                return NULL;
+
+        *u = (Userdata) {
+                .userdata = userdata,
+                .bus = bus,
+                .event = sd_bus_get_event(bus),
+                .remaining = remaining,
+                .target_path = target_path,
+                .target_id = target_id,
+        };
+        return TAKE_PTR(u);
+}
+
+static int ensure_targets(sd_bus *bus, char **argv, char ***ret_targets) {
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_strv_free_ char **targets = NULL;
+        int r;
+
+        assert(bus);
+        assert(ret_targets);
+
+        if (strv_isempty(argv)) {
+                const char *class, *name, *path;
+
+                r = bus_call_method(bus, bus_sysupdate_mgr, "ListTargets", &error, &reply, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to call ListTargets: %s", bus_error_message(&error, r));
+
+                r = sd_bus_message_enter_container(reply, 'a', "(sso)");
+                if (r < 0)
+                        return r;
+
+                while ((r = sd_bus_message_read(reply, "(sso)", &class, &name, &path)) > 0) {
+                        _cleanup_free_ char *id = NULL;
+
+                        if (streq(class, "host"))
+                                id = strdup(".host");
+                        else
+                                id = strjoin(class, ":", name);
+                        if (!id)
+                                return log_oom();
+
+                        r = strv_consume(&targets, TAKE_PTR(id));
+                        if (r < 0)
+                                return r;
+                }
+                if (r < 0)
+                        return r;
+
+                r = sd_bus_message_exit_container(reply);
+                if (r < 0)
+                        return r;
+
+        } else {
+                r = strv_extend_strv(&targets, argv, true);
+                if (r < 0)
+                        return r;
```

Comment:
move these two lines out of the else block, and share it with the other if branch. you can save 3 lines of code!